File: Python/Binary_Search.py
# Binary Search = An efficent way to search in a sorted list
def binary(arr,low,high,x):
    while low < high:
        mid = low + (high-1) // 2
        if(arr[mid] == x):
            return mid
        elif(arr[mid] <x):
            low = mid + 1
        else:
            high = mid - 1
    return -1

print('Enter a list of number')
arr = list(map(int,input().split()))
n=len(arr)

arr.sort()
x =int(input("Enter the number to find\n"))
result = binary(arr,0,len(arr)-1,x)
if(result == -1):
    print("Element not found")
else:
    print("Element found at ",result+1," position")

'''
Time complexity ==> O(logn)
Space Complexity ==> O(1)

I/O ==>

Enter a list of number
234 567 321 345 123 110 767 111
Enter the number to find
767
Element found at  8  position
'''

File: Go/BinarySearch.go
package main

import (
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

// Implementation of binary search without using recursion
func search(array []int, elementToFind int) int {
	low := 0
	high := len(array) - 1

	for low <= high {
		// Find the middle index of array
		index := int((low + high) >> 1)

		// If find the element, return the position
		if array[index] == elementToFind {
			return index
		} else if array[index] < elementToFind {
			// The target value is on the upper portion of the array
			low = index + 1
		} else {
			// The target value is on the lower portion of the array
			high = index - 1
		}
	}

	// The searched item was not found
	return -1
}

func main() {
	var array []int
	var elementToFind int

	// Converts all input arguments to integers and adds them to the list to be sorted
	for _, value := range os.Args[1:] {
		number, err := strconv.Atoi(value)

		if err != nil {
			log.Fatal(err)
		}

		array = append(array, number)
	}

	// Sort array of ints in increasing order
	sort.Ints(array)

	// Asks the user to enter the number they want to find within the array
	fmt.Print("Enter the number to find: ")
	fmt.Scanf("%d", &elementToFind)

	fmt.Println(fmt.Sprintf("Element %d was found at position %d", elementToFind, search(array, elementToFind)))
}

/*
I => Array: 12 19 20 33 44 68 79 80
	 Element to find: 68
O => 5

Time complexity: O(log n)
Space complexity: O(1)
*/
